storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -241,64 +255,54 @@ static DynamicState prepareForNewAssignmentNoWorkersRunning(DynamicState dynamic
-                                                                                 staticState.changingCallback);
+        //Skip special case if `storm kill_workers` is already invoked
+        Boolean isDead = dynamicState.container.areAllProcessesDead();
+        if (!isDead) {
+            //TODO: Is this necessary?
+            if (reason == KillReason.ASSIGNMENT_CHANGED || reason == KillReason.BLOB_CHANGED) {
+                staticState.iSupervisor.killedWorker(staticState.port);
+            }
+            dynamicState.container.kill();
         }
-        dynamicState = drainAllChangingBlobs(dynamicState);
-        Time.sleep(staticState.killSleepMs);
-        return dynamicState.withPendingLocalization(dynamicState.newAssignment, pendingDownload).withState(MachineState.KILL);
-    }
-
-    /**
-     * Kill the current container, and wait go to the state to inform the localizer that we are ready to go.
-     * PRECONDITION: container != null
-     * @param dynamicState current state
-     * @param staticState static data
-     * @return the next state
-     */
-    private static DynamicState killContainerForChangedBlobs(DynamicState dynamicState, StaticState staticState) throws Exception {
-        assert (dynamicState.container != null);
-
-        staticState.iSupervisor.killedWorker(staticState.port);
-        dynamicState.container.kill();
+        numWorkersKilledFor.get(reason).mark();
 
-        Time.sleep(staticState.killSleepMs);
-        return dynamicState.withState(MachineState.KILL_BLOB_UPDATE);
-    }
-
-    /**
-     * Kill the current container and relaunch it.  (Something odd happened)
-     * PRECONDITION: container != null
-     * @param dynamicState current state
-     * @param staticState static data
-     * @return the next state
-     * @throws Exception
-     */
-    static DynamicState killAndRelaunchContainer(DynamicState dynamicState, StaticState staticState) throws Exception {
-        assert (dynamicState.container != null);
+        DynamicState next;
+        switch (reason) {
+            case ASSIGNMENT_CHANGED:
+                Future<Void> pendingDownload = null;
+                if (dynamicState.newAssignment != null) {
+                    pendingDownload = staticState.localizer.requestDownloadTopologyBlobs(
+                            dynamicState.newAssignment, staticState.port, staticState.changingCallback);
+                }
+                dynamicState = drainAllChangingBlobs(dynamicState);
+                next = dynamicState.withState(MachineState.KILL)
+                        .withPendingLocalization(dynamicState.newAssignment, pendingDownload);
+                break;
+
+            case BLOB_CHANGED:
+                next = dynamicState.withState(MachineState.KILL_BLOB_UPDATE);
+                break;
+
+            case PROCESS_EXIT: case MEMORY_VIOLATION:
+            case HB_TIMEOUT: case HB_NULL:
```

Comment:
nit: can we put each of these on their own line.  It just feels cleaner to me that way.